include/kframework/lemmas/bytes-simplification.k

```diff
@@ -68,71 +68,168 @@ module BYTES-SIMPLIFICATION [symbolic]
+      #range(B1, S1, W1) +Bytes #range(B1, S2, W2) +Bytes B2 => #range(B1, S1, W1 +Int W2) +Bytes B2
+      requires S2 ==Int S1 +Int W1
+       andBool 0 <=Int S1 andBool 0 <=Int W1 andBool 0 <=Int S2 andBool 0 <=Int W2
+      [simplification]
 
-    rule #range( M [ N1 := BUF1:Bytes ] [ N2 := BUF2:Bytes ], L, K ) => #range( M [ N2 := BUF2 ], L, K )
-        requires 0 <Int K andBool N1 +Int lengthBytes(BUF1) <=Int L [simplification]
+    rule [range-reform-base]:
+      #range(B:Bytes, S:Int, minInt(W1:Int, W2:Int)) +Bytes #buf(maxInt(0, W3:Int), 0) => #range (B, S, W1)
+      requires W2 ==Int lengthBytes(B) -Int S
+       andBool W3 ==Int S +Int W1 -Int lengthBytes(B)
+       andBool 0 <=Int S andBool 0 <=Int W1 andBool 0 <=Int W2
+      [simplification]
+
+    rule [range-reform-concat]:
+      #range(B1:Bytes, S:Int, minInt(W1:Int, W2:Int)) +Bytes #buf(maxInt(0, W3:Int), 0) +Bytes B2:Bytes => #range (B1, S, W1) +Bytes B2
+      requires W2 ==Int lengthBytes(B1) -Int S
+       andBool W3 ==Int S +Int W1 -Int lengthBytes(B1)
+       andBool 0 <=Int S andBool 0 <=Int W1 andBool 0 <=Int W2
+      [simplification]
 
-    rule #range(M, N1 +Int N2, K) => #range(#range(M, N2, lengthBytes(M) -Int N2), N1, K)
-      requires 0 <=Int N1 andBool 0 <=Int N2
-      [simplification, concrete(N2), concrete(M)]
+    rule [range-buf-zero-1]:
+      #range(#buf(32, X:Int), 0, W:Int) => #buf(W, 0)
+      requires 0 <=Int W andBool W <=Int 32
+       andBool #rangeUInt (8 *Int (32 -Int W), X)
+      [simplification, concrete(W)]
 
-    rule #range(A +Bytes _B, START, LENGTH) => #range(A, START, LENGTH)
-      requires START +Int LENGTH <=Int lengthBytes(A) [simplification]
+    rule [range-buf-zero-2-symb]:
+      #range(#buf(W:Int, 0), S1:Int, W1:Int) => #buf(W1, 0)
+      requires 0 <=Int W andBool 0 <=Int S1 andBool 0 <=Int W1
+      [simplification]
 
-    rule #range(A +Bytes B, START, LENGTH) => #range(B, START -Int lengthBytes(A), LENGTH)
-      requires lengthBytes(A) <=Int START [simplification]
+    rule [range-buf-zero-2-conc]:
+      #range(B, S1:Int, W1:Int) => #buf(W1, 0)
+      requires 0 <=Int S1 andBool 0 <=Int W1
+       andBool B ==K #buf(lengthBytes(B), 0)
+      [concrete(B), simplification]
+
+    rule [range-memUpdate-before]:
+      #range(B1:Bytes [ S1:Int := B2:Bytes ], S2:Int, W2:Int) =>
+        #let W = minInt(W2, S1 -Int S2) #in
+          #range(B1, S2, W) +Bytes #range(B1 [ S1 := B2 ], S1, W2 -Int W)
+      requires S2 <Int S1
+      [simplification]
 
-    rule #range(BA:Bytes, 0, LEN) => BA requires lengthBytes(BA) ==Int LEN [simplification]
+    rule [range-memUpdate-in-between]:
+      #range(B1:Bytes [ S1:Int := B2:Bytes ], S2:Int, W2:Int) =>
+        #let S = S2 -Int S1 #in
+        #let W = minInt(W2, lengthBytes(B2) -Int S) #in
+          #range(B2, S, W) +Bytes #range(B1, S1 +Int lengthBytes(B2), W2 -Int W)
+      requires S1 <=Int S2 andBool S2 <=Int S1 +Int lengthBytes(B2)
+      [simplification]
```

Comment:
Do we need here additionally that `0 <=Int S1`?

Reply:
Very probably, yes.